media/201903-report-chalmers.md

```diff
@@ -0,0 +1,750 @@
+
+
+
+The \K framework lets one define programming languages by giving their syntax
+and semantics. The vision of \K is to provide a language-independent framework
+for obtaining not only parsers, interpreters and compilers, but also more
+non-standard tools like documentation, program verifiers, and model
+checkers. Once a language is definied, all these tools should follow.
+Perhaps the most interesting tool that comes out of a \K specification is the
+Reachability Logic Prover [@stefanescu-park-yuwen-li-rosu-reachability-prover].
+
+![\K Tooling Overview](media/img/k-overview.png)
+
+Several large and popular programming languages have been specified in \K, among
+them Java 1.4 [@bogdanas-rosu-k-java], C11 [@hathhorn-ellison-rosu-k-c], and
+JavaScript [@park-stefanescu-rosu-k-js]. There are ongoing projects to specfy
+LLVM[^llvm] and X86 assemby[^x86].
+
+[^llvm]: <https://github.com/kframework/llvm-semantics>
+[^x86]: <https://github.com/kframework/X86-64-semantics>
+
+A language specification in \K consists of 3 things:
+
+* the concrete syntax,
+* a configuration, which is the exectuion state,
+* operational semantics as rewrite rules on the configuration.
+
+## \K Specifications: Syntax ##
+
+The concrete syntax is built using EBNF style. As a running example, we use a small, imperative language.
+
+```k
+    syntax Exp ::= Int | Id | "(" Exp ")" [bracket]
+                 | Exp "*" Exp
+                 > Exp "+" Exp // looser binding
+
+    syntax Stmt ::= Id ":=" Exp
+                  | Stmt ";" Stmt
+                  | "return" Exp
+```
+
+This would allow correctly parsing programs like:
+
+```imp
+    a := 3 * 2;
+    b := 2 * a + 5;
+    return b
+```
+
+## \K Specifications: Configuration ##
+
+The configuration gives the structure of the execution state.
+For example, a simple imperative language might have the following
+configuration.
+
+```k
+    configuration <k>     $PGM:Program </k>
+                  <env>   .Map         </env>
+                  <store> .Map         </store>
+```
+
+-   `<k>` will contain the initial parsed program
+-   `<env>` contains bindings of variable names to store locations
+-   `<store>` contains bindings of store locations to integers
+
+
+Note that the parsed program gets put into the `<k>` cell. Thus, the program is
+part of the execution state, and can be reduced or extended as part of rewrites.
+
+## \K Specifications: Transition Rules ##
+
+The actual operational semantics of a \K program is written as rewrite rules. For example, we might define the following semantic rules for our imperative language.
+
+A transition rule takes the form
+
+```k
+<cellA> X => Y </cellA>
+<cellB> Z => W </cellB>
+etc.
+```
+
+where the arrow represents a rewrite: the value to the left is rewritten to the
+value on the right. The above rule says that "if `<cellA>` contains `X` and `<cellB>` contains `Z`, change the contents of `<cellA>` to `Y` and the contents of `<cellB>` to `W`."
```

Comment:
Emphasize that the rewrite on both cells happens simultaneously, and that the rewrite being contained inside the cell is only a syntactic convenience.

Reply:
Done